File: src/dto/ColorDTO.ts
import { BaseDTO } from './BaseDTO';
import { Color } from '../model';

export class ColorDTO extends BaseDTO {
  red: number;
  green: number;
  blue: number;
  alpha?: number;

  constructor(red: number, green: number, blue: number, alpha?: number) {
    super();
    this.red = red;
    this.green = green;
    this.blue = blue;
    this.alpha = alpha;
  }

  static fromDataModel(color: Color) {
    return new ColorDTO(color.red, color.green, color.blue, color.alpha);
  }

  static fromJSON(json: string) {
    const parsedJson = JSON.parse(json);
    return new ColorDTO(
      parsedJson.red,
      parsedJson.green,
      parsedJson.blue,
      parsedJson.alpha
    );
  }

  toDataModel(): Color {
    return new Color(this.red, this.green, this.blue, this.alpha);
  }

  toJSON() {
    return {
      red: this.red,
      green: this.green,
      blue: this.blue,
      alpha: this.alpha,
    };
  }
}

File: android/src/main/java/com/paintboxreactnative/dto/ColorDTO.kt
package com.paintboxreactnative.dto

import android.annotation.TargetApi
import android.graphics.Color
import android.os.Build
import androidx.annotation.RequiresApi

data class ColorDTO(
  val red: Double,
  val green: Double,
  val blue: Double,
  val alpha: Double = 255.0,
) {
  companion object {
    @TargetApi(Build.VERSION_CODES.O)
    @RequiresApi(Build.VERSION_CODES.O)
    fun fromDataModel(color: Color): ColorDTO {
      return ColorDTO(
        red = color.red().toDouble(),
        green = color.green().toDouble(),
        blue = color.blue().toDouble(),
        alpha = color.alpha().toDouble()
      )
    }
  }

  @TargetApi(Build.VERSION_CODES.O)
  @RequiresApi(Build.VERSION_CODES.O)
  fun toDataModel(): Color {
    return Color.valueOf(
      (red / 255.0).toFloat(),
      (green / 255.0).toFloat(),
      (blue.toFloat() / 255.0).toFloat(),
      (alpha.toFloat() / 255.0).toFloat(),
    )
  }
}
